man/man1/ptest.1
.\"
.\" $Id: root-config.1,v 1.4 2005/03/21 21:42:21 rdm Exp $
.\"
.TH PTEST 1 "Version 5" "ROOT"
.\" NAME should be all caps, SECTION should be 1-8, maybe w/ subsection
.\" other parms are allowed: see man(7), man(1)
.SH NAME
ptest \- xproofd test utility
.SH SYNOPSIS
.B ptest
.I "[options]"
.SH "DESCRIPTION"
This manual page documents the
.BR ptest
program, an utility whose purpose is to test the status of an xproofd daemon running on the local machine.
The target is to run
.B ptest
inside a monitor infrastructure, for example
.B monit
(see
.UR http://mmonit.com/monit/
\fIhttp://mmonit.com/monit/\fP
), capable to take some decisions and to do some actions following the result of
.B ptest.


.SH OPTIONS 
.TP 
.B \-\-help, \-h
Gives a short list of options avaliable, and exit
.TP
.B \-t\fP \fR<test>\fP 
type of test to be run:
    0       ping the daemon
    1       ping the daemon and check connection for default user
    2       ping the daemon and check connection for the default user and all recent users

The default user is the one specified in the URL (e.g. 'user' for user@localhost); if no explicit specification is found in the URL, the user under which
.B ptest
is run is used; it can be also passed via the environment variable
.B PTEST_TEST.
.TP
.B \-u\fP \fR<url>\fP
URL where the xproofd under test responds; it can be also passed via the environment variable
.B PTEST_URL
(default is 'localhost:1093').
.TP
.B \-d\fP \fR<sbdir>\fP
sandbox directory used to find out the users of the facility when test is 2; it can be also passed via the environment variable
.B PTEST_SBOXDIR
(default is '/tmp/proofbox').
.TP
.B \-s\fP \fR<span>\fP
define the time interval, in minutes, to define 'recent' users when test is 2: only users who connected within this interval are checked; use -1 for infinite;
it can be also passed via the environment variable
.B PTEST_TIMESPAN
(default is -1).
.TP
.B \-l\fP \fR<logfile>\fP
log file if not screen; deleted if the required test fails unless '-k' is specified (see below);it can be also passed via the environment variable
.B PTEST_LOGFILE.
.TP
.B \-k
keep log file at path given via '-l' in all cases; it can be also passed via the environment variable
.B PTEST_KEEP
(default is 'no keep').
.SH RETURN
.B ptest
returns 0 on success and 1 in case of test failure. Errors are printed on the screen or saved into the specified log file.

.SH FILES
The file
.B ptest_popen_file
in the temporary directory (typically /tmp) is used to parse the output of
.B TProof::Open().
The file is overwritten by each new attempt; after a run of
.B ptest
it contains the result of the last session open attempt.

.SH NOTES
The connection attempt is done in
.B 'masteronly'
mode. For daemons running in
.B 'worker'
mode the attempt will fail with a well defined error message ('Server not allowed to be top master');
since this indicates that the daemon is anyhow responsive, a positive detection of such a pattern in the error message
is interpreted as the attempt having being successful.

.SH "ORIGINAL AUTHORS"
The ROOT team:
.RS
\fBRene Brun\fR and \fBFons Rademakers\fR
.RE
More information can be found at the \fBROOT\fR website:
.UR http://root.cern.ch
\fIhttp://root.cern.ch\fB
.SH "COPYRIGHT"
This library is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as
published by the Free Software Foundation; either version 2.1 of the
License, or (at your option) any later version.
.P
This library is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.
.P
You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
.SH AUTHOR 
This manual page was written by Gerardo Ganis
<gerardo.ganis@cern.ch>, for the GNU/Linux system (but may be used by
others). 
.\"
.\" EOF
.\"


main/src/ptest.cxx
// @(#)root/main:$Id$
// Author: G. Ganis 2012

/*************************************************************************
 * Copyright (C) 1995-2000, Rene Brun and Fons Rademakers.               *
 * All rights reserved.                                                  *
 *                                                                       *
 * For the licensing terms see $ROOTSYS/LICENSE.                         *
 * For the list of contributors see $ROOTSYS/README/CREDITS.             *
 *************************************************************************/

//////////////////////////////////////////////////////////////////////////
//                                                                      //
// PTest                                                                //
//                                                                      //
// Program used to test existence and status of a proof daemon          //
//                                                                      //
//////////////////////////////////////////////////////////////////////////
#include <stdio.h>
#include <errno.h>

#include <sys/time.h>
#include <sys/resource.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>

#include "TSystem.h"
#include "TProof.h"
#include "TString.h"

//______________________________________________________________________________
// Prototypes
int xpd_ping(const char *host, int port);
int getsocket(struct hostent *h, int);
void printhelp();
int recvn(int sock, void *buffer, int length);
int sendn(int sock, const void *buffer, int length);
int proof_open(const char *master);
int parse_args(int argc, char **argv,
               TString &url, TString &sboxdir, time_t &span, int &test,
               TString &logfile, bool &keep);
void printhelp();

//______________________________________________________________________________
// The client request structure
typedef struct {
   int first;
   int second;
   int third;
   int fourth;
   int fifth;
} clnt_HS_t;

//______________________________________________________________________________
// The body received after the first handshake's header
typedef struct {
   int msglen;
   int protover;
   int msgval;
} srv_HS_t;

//______________________________________________________________________________
int main(int argc, char **argv)
{
   //  PROOF daemon test program.
   //  Syntax
   //          ptest  <url> <sandbox_dir> <time_span>
   // 
   //          <url>            URL to test; default 'localhost:1093'
   //          <sandbox_dir>    directory with users sandboxes; used to find out
   //                           users to check connection; default '/tmp/proofbox'
   //          <time_span>      check only users whose latest activity was within
   //                           'time_span' minutes; use -1 to check all users;
   //                           default -1.
   //
   //  Exits 0 on success, 1 on error

   TString url, sboxdir, logfile;
   time_t span = -1;
   int test = 0;
   bool keep = 0;
   int rc = parse_args(argc, argv, url, sboxdir, span, test, logfile, keep);
   if (rc < 0) {
      fprintf(stderr, "ptest: parse_args failure\n");
      gSystem->Exit(1);
   } else if (rc > 0) {
      gSystem->Exit(0);
   }
   
   rc = 0;
   
   // Set up log file if required
   RedirectHandle_t redirH;
   if (!logfile.IsNull()) {
      gSystem->RedirectOutput(logfile, "w", &redirH);
   }

   // Check sandbox dir
   FileStat_t fst;
   if (gSystem->GetPathInfo(sboxdir, fst) != 0) {
      fprintf(stderr, "ptest: stat failure for '%s'\n", sboxdir.Data());
      rc = 1;
   }
   if (rc == 0 && !R_ISDIR(fst.fMode)) {
      fprintf(stderr, "ptest: '%s' is not a directory\n", sboxdir.Data());
      rc = 1;
   }
    
   // Setup URL
   TUrl u(url.Data());
   TString defusr = u.GetUser();
   if (defusr.IsNull()) {
      UserGroup_t *pw = gSystem->GetUserInfo();
      if (pw) {
         defusr = pw->fUser;
         delete pw;
      }
   }
   if (!url.BeginsWith(u.GetProtocol())) {
      if (u.GetPort() == 80 && !strcmp(u.GetProtocol(), "http")) u.SetPort(1093);
      u.SetProtocol("proof");
   }

   // Do ping 
   if (rc == 0)
      if ((rc = xpd_ping(u.GetHost(), u.GetPort())) != 0)
         fprintf(stderr, "ptest: failure pinging '%s'\n", url.Data());

   if (test > 0) {
      // Do TProof::Open in "masteronly" mode
      if (rc == 0) {
         if (!logfile.IsNull()) gSystem->RedirectOutput(0, 0, &redirH);
         rc = proof_open(u.GetUrl());
         if (!logfile.IsNull()) gSystem->RedirectOutput(logfile, "a", &redirH);
         if (rc != 0)
            fprintf(stderr, "ptest: TProof::Open failure for default user '%s'\n", u.GetUrl());
      }

      if (test > 1) {
         if (rc == 0) {

            // Scan sandbox dir
            time_t now = time(0);
            void *dirp = gSystem->OpenDirectory(sboxdir.Data());
            const char *ent = 0;
            TString dent;
            while ((ent = gSystem->GetDirEntry(dirp))) {
               if (!strcmp(ent, "..") || !strcmp(ent, ".")) continue;
               if (defusr == ent) continue;
               FileStat_t st;
               if (span > 0) {
                  dent.Form("%s/%s", sboxdir.Data(), ent);
                  if (gSystem->GetPathInfo(dent, st) != 0) {
                     fprintf(stderr, "ptest: stat failure for '%s'\n", dent.Data());
                     rc = 1;
                     break;
                  }
               }
               if (span < 0 || st.fMtime > (now - span)) {
                  u.SetUser(ent);
                  fprintf(stderr, "proof_open: url: '%s'\n", u.GetUrl());
                  if (!logfile.IsNull()) gSystem->RedirectOutput(0, 0, &redirH);
                  rc = proof_open(u.GetUrl());
                  if (!logfile.IsNull()) gSystem->RedirectOutput(logfile, "a", &redirH);
                  if (rc != 0) {
                     fprintf(stderr, "ptest: failure scanning sandbox dir '%s'\n", sboxdir.Data());
                     break;
                  }
               }
            }
            gSystem->FreeDirectory(dirp);
         }
      }
   }
   if (!logfile.IsNull()) {
      gSystem->RedirectOutput(0, 0, &redirH);
      if (rc == 0 && !keep) gSystem->Unlink(logfile);
   }
   gSystem->Exit(rc);
}

// Auxilliary functions

//______________________________________________________________________________
int parse_args(int argc, char **argv,
               TString &url, TString &sboxdir, time_t &span, int &test,
               TString &logfile, bool &keep)
{
   // Extract control info from arguments
   
   url = "localhost:1093";
   sboxdir = "/tmp/proofbox";
   span = -1;
   test = 0;
   logfile= "";
   keep = 0;

   // Check environment settings first
   if (getenv("PTEST_URL")) {
      url = getenv("PTEST_URL");
   }
   if (getenv("PTEST_SBOXDIR")) {
      sboxdir = getenv("PTEST_SBOXDIR");
   }
   if (getenv("PTEST_TIMESPAN")) {
      errno = 0;
      long xspan = strtol(getenv("PTEST_TIMESPAN"), 0, 10);
      if (errno == 0) span = (time_t) xspan;
   }
   if (getenv("PTEST_TEST")) {
      errno = 0;
      long xtest = strtol(getenv("PTEST_TEST"), 0, 10);
      if (errno == 0 && xtest >= 0 && xtest <= 2) sscanf(getenv("PTEST_TEST"), "%d", &test);
   }
   if (getenv("PTEST_LOGFILE")) {
      logfile = getenv("PTEST_LOGFILE");
   }
   if (getenv("PTEST_KEEP")) {
      keep = 1;
   }

   // -u url -d sboxdir -s span -t test
   for (int i = 1; i < argc; i++) {
      if (argv[i]) {
         if (!strcmp(argv[i], "-h") || !strcmp(argv[i], "--help")) {
            printhelp();
            return 1;
         } else if (!strcmp(argv[i], "-u")) {
            if (argv[++i]) {
               url = argv[i];
            } else {
               fprintf(stderr, "parse_args: '-u' requires the specification of the URL!\n");
               printhelp();
               return -1;
            }
         } else if (!strcmp(argv[i], "-d")) {
            if (argv[++i]) {
               sboxdir = argv[i];
            } else {
               fprintf(stderr, "parse_args: '-d' requires the specification of the sandbox directory!\n");
               printhelp();
               return -1;
            }
         } else if (!strcmp(argv[i], "-l")) {
            if (argv[++i]) {
               logfile = argv[i];
            } else {
               fprintf(stderr, "parse_args: '-l' requires the specification of the logfile!\n");
               printhelp();
               return -1;
            }
         } else if (!strcmp(argv[i], "-k")) {
            keep = 1;
         } else if (!strcmp(argv[i], "-s")) {
            if (argv[++i]) {
               errno = 0;
               long xspan = strtol(argv[i], 0, 10);
               if (errno == 0 && xspan > 0) {
                  span = (time_t) xspan;
               } else {
                  fprintf(stderr, "parse_args: time span must be a positive integer! (argv: %s, xspan: %ld)\n",
                                  argv[i], xspan);
                  printhelp();
                  return -1;
               }
            } else {
               fprintf(stderr, "parse_args: '-s' requires the specification of the time span!\n");
               printhelp();
               return -1;
            }
         } else if (!strcmp(argv[i], "-t")) {
            if (argv[++i]) {
               errno = 0;
               long xtest = strtol(argv[i], 0, 10);
               if (errno == 0 && xtest >= 0 && xtest <= 2) {
                  sscanf(argv[i], "%d", &test);
               } else {
                  fprintf(stderr, "parse_args: test must be an integer in [0,2]! (argv: %s, xtest: %ld)\n",
                                  argv[i], xtest);
                  printhelp();
                  return -1;
               }
            } else {
               fprintf(stderr, "parse_args: '-t' requires the specification of the test!\n");
               printhelp();
               return -1;
            }
         }
      }
   }
   // Done
   return 0;
}

//______________________________________________________________________________
void printhelp()
{
   // Help function
   fprintf(stderr, "\n");
   fprintf(stderr, "   ptest: test xproofd service on (remote) host\n");
   fprintf(stderr, "\n");
   fprintf(stderr, "   Syntax:\n");
   fprintf(stderr, "            ptest [-u url] [-t test] [-d sbdir] [-s span] [-l logfile] [-k] [-h|--help]\n");
   fprintf(stderr, "\n");
   fprintf(stderr, "   url:     URL where the xproofd under test responds [localhost:1093]\n");
   fprintf(stderr, "   test:    type of test [0]\n");
   fprintf(stderr, "                 0   ping\n");
   fprintf(stderr, "                 1   '0' + check connection for default user\n");
   fprintf(stderr, "                 2   '1' + check connection for all recent users\n");
   fprintf(stderr, "   sbdir:   sandbox directory used to find out the users of the facility\n");
   fprintf(stderr, "            when test is 2 ['/tmp/proofbox']\n");
   fprintf(stderr, "   span:    define the time interval to define 'recent' users when test\n");
   fprintf(stderr, "            is 2: only users who connected within this interval are checked;\n");
   fprintf(stderr, "            use -1 for infinite [-1]\n");
   fprintf(stderr, "   logfile: log file if not screen; deleted if the test fails unless '-k' is\n");
   fprintf(stderr, "            specified (see below) [terminal]\n");
   fprintf(stderr, "   -k:      keep log file at path given via '-l' in all cases [do not keep]\n");
   fprintf(stderr, "   -h, --help: print this screen\n");
   fprintf(stderr, "\n");
   fprintf(stderr, "   Return:  0   the required test succeeded\n");
   fprintf(stderr, "            1   the test failed\n");
   fprintf(stderr, "\n");
}

//______________________________________________________________________________
int proof_open(const char *master)
{
   // Test proof open.
   // Return 0 on success, 1 on failure
   int rc = 0;
   RedirectHandle_t rh;
   TString popenfile = TString::Format("%s/ptest_popen_file", gSystem->TempDirectory());
   gSystem->RedirectOutput(popenfile, "w", &rh);
   TProof *p = TProof::Open(master, "masteronly");
   gSystem->RedirectOutput(0, 0, &rh);
   if (!p || (p && !p->IsValid())) {
      TMacro m;
      m.ReadFile(popenfile);
      if (!m.GetLineWith("Server not allowed to be top master")) rc = 1;
   }
   if (p) delete p;
   return rc;
}

//______________________________________________________________________________
int xpd_ping(const char *host, int port)
{
   // Non-blocking check for a PROOF service at 'url'
   // Return
   //        0 if a XProofd daemon is listening at 'url'
   //        1 otherwise

   // Check arguments
   if (!host || (host && strlen(host) <= 0)) {
      fprintf(stderr,"xpd_ping: host must be given!\n");
      return 1;
   }

   struct hostent *h = gethostbyname(host);
   if (!h) {
      fprintf(stderr,"xpd_ping: unknown host '%s'\n", host);
      return 1;
   }

   // Get socket and listen to it
   int sd = getsocket(h,port);
   if (sd == -1) {
      fprintf(stderr,"xpd_ping: problems creating socket ... exit\n");
      return 1;
   }

   // Send the first bytes
   clnt_HS_t initHS;
   memset(&initHS, 0, sizeof(initHS));
   int len = sizeof(initHS);
   initHS.third  = (int)htonl((int)1);
   if (sendn(sd, &initHS, len) != len) {
      fprintf(stderr,"xpd_ping: problems sending first set of handshake bytes\n");
      return 1;
   }
   
   // These 8 bytes are need by 'rootd/proofd' and discarded by XRD/XPD
   int dum[2];
   dum[0] = (int)htonl((int)4);
   dum[1] = (int)htonl((int)2012);
   if (sendn(sd, &dum[0], sizeof(dum)) != sizeof(dum)) {
      fprintf(stderr,"xpd_ping: problems sending second set of handshake bytes\n");
      return 1;
   }
   
   // Read first server response
   int type;
   len = sizeof(type);
   int nr = 0;
   if ((nr = recvn(sd, &type, len)) != len) { // 4 bytes
      fprintf(stderr, "xpd_ping: 1st: wrong number of bytes read: %d (expected: %d)\n",
                      nr, len);
      return 1;
   }
   
   // To host byte order
   type = ntohl(type);
   // Check if the server is the eXtended proofd
   if (type == 0) {
      srv_HS_t xbody;
      len = sizeof(xbody);
      if ((nr = recvn(sd, &xbody, len)) != len) { // 12(4+4+4) bytes
         fprintf(stderr, "xpd_ping: 2nd: wrong number of bytes read: %d (expected: %d)\n",
                         nr, len);
         return 1;
      }
      xbody.protover = ntohl(xbody.protover);
      xbody.msgval = ntohl(xbody.msglen);
      xbody.msglen = ntohl(xbody.msgval);

   } else if (type == 8) {
      // Standard proofd
      fprintf(stderr, "xpd_ping: server is PROOFD\n");
      return 1;
   } else {
      // We don't know the server type
      fprintf(stderr, "xpd_ping: unknown server type: %d\n", type);
      return 1;
   }
   // Cleanup
   close(sd);
   // Done
   return 0;
}

//______________________________________________________________________________
int getsocket(struct hostent *h, int port)
{
   int sd, rc;
   struct sockaddr_in localAddr, servAddr;

   servAddr.sin_family = h->h_addrtype;
   memcpy((char *) &servAddr.sin_addr.s_addr, h->h_addr_list[0], h->h_length);
   servAddr.sin_port = htons(port);

   /* create socket */
   sd = socket(AF_INET, SOCK_STREAM, 0);
   if(sd < 0) {
      perror("cannot open socket ");
      return -1;
   }

   /* bind any port number */
   localAddr.sin_family = AF_INET;
   localAddr.sin_addr.s_addr = htonl(INADDR_ANY);
   localAddr.sin_port = htons(0);
  
   rc = bind(sd, (struct sockaddr *) &localAddr, sizeof(localAddr));
   if(rc < 0) {
      perror("error ");
      return -1;
   }
                                
   /* connect to server */
   rc = connect(sd, (struct sockaddr *) &servAddr, sizeof(servAddr));
   if(rc < 0) {
      perror("cannot connect ");
      return -1;
   }

   return sd;

}

//______________________________________________________________________________
int sendn(int sock, const void *buffer, int length)
{
   // Send exactly length bytes from buffer.

   if (sock < 0) return -1;

   int n, nsent = 0;
   const char *buf = (const char *)buffer;

   for (n = 0; n < length; n += nsent) {
      if ((nsent = send(sock, buf+n, length-n, 0)) <= 0) {
         perror("problems sending ");
         return nsent;
      }
   }

   return n;
}

//______________________________________________________________________________
int recvn(int sock, void *buffer, int length)
{
   // Receive exactly length bytes into buffer. Returns number of bytes
   // received. Returns -1 in case of error.

   if (sock < 0) return -1;

   int n, nrecv = 0;
   char *buf = (char *)buffer;

   for (n = 0; n < length; n += nrecv) {
      while ((nrecv = recv(sock, buf+n, length-n, 0)) == -1 && errno == EINTR)
         errno = 0;   // probably a SIGCLD that was caught
      if (nrecv < 0) {
         perror("problems receiving ");
         return nrecv;
      } else if (nrecv == 0)
         break;         // EOF
   }

   return n;
}
